packages/block-editor/src/hooks/align.js

```diff
@@ -207,6 +212,130 @@ export const withDataAlign = createHigherOrderComponent(
 		return <BlockListBlock { ...props } wrapperProps={ wrapperProps } />;
 	}
 );
+export function AlignmentVisualizer( {
+	value = 'none',
+	allowedAlignments,
+	clientId,
+} ) {
+	const layout = useLayout();
+	const blockName = useSelect(
+		( select ) => select( blockEditorStore ).getBlockName( clientId ),
+		[ clientId ]
+	);
+
+	const [ editorWidth, setEditorWidth ] = useState( 0 );
+	const ref = useRefEffect( ( node ) => {
+		const { ownerDocument } = node;
+		const { defaultView } = ownerDocument;
+		const editorWrapperElement = ownerDocument.querySelector(
+			'.editor-styles-wrapper'
+		);
+		const updateWidth = () =>
+			setEditorWidth( editorWrapperElement.offsetWidth );
+		updateWidth();
+
+		const resizeObserver = defaultView.ResizeObserver
+			? new defaultView.ResizeObserver( updateWidth )
+			: undefined;
+		resizeObserver?.observe( editorWrapperElement );
+
+		return () => {
+			resizeObserver?.disconnect();
+		};
+	}, [] );
+
+	const blockAllowedAlignments = getValidAlignments(
+		getBlockSupport( blockName, 'align' ),
+		hasBlockSupport( blockName, 'alignWide', true )
+	);
+
+	// Allow override of `blockAllowedAlignments`. The image block doesn't use
+	// alignment block supports, so this allows the image block to directly
+	// declare what it supports.
+	const availableAlignments = useAvailableAlignments(
+		allowedAlignments ?? blockAllowedAlignments
+	);
+
+	const alignments = useMemo( () => {
+		return [ 'none', 'wide', 'full' ]
+			.map( ( name ) => {
+				if ( name === 'none' ) {
+					return {
+						name,
+						label: __( 'Content width' ),
+						style: {
+							width: layout.contentSize,
+						},
+					};
+				}
+				if ( name === 'wide' ) {
+					return {
+						name,
+						label: __( 'Wide width' ),
+						style: {
+							width: layout.wideSize,
+						},
+					};
+				}
+				if ( name === 'full' ) {
+					return {
+						name,
+						label: __( 'Full width' ),
+						style: {
+							width: '100%',
+						},
+					};
+				}
+				return null;
+			} )
+			.filter( ( alignment ) => alignment !== null );
+	}, [ editorWidth, availableAlignments, layout ] );
+
+	const wrapperStyle = useMemo( () => {
+		const currentAlignment = alignments.find(
+			( alignment ) => alignment.name === value
+		);
+		const offset = `calc( ( ${ editorWidth }px - ${ currentAlignment.style.width } ) / -2 )`;
```

Comment:
This calculation may be too brittle. I think it would be better if there were a way to override the `BlockPopover` anchor rect. The popover could anchor to the editor's left/right, but the block's top/bottom.

Potentially it may be best to not use `BlockPopover` for this.